### ISSUE TITLE
Can't use filter through h5repack

### ISSUE THREAD

**Issue by estan** [Blosc/hdf5-blosc#13]
I'm trying to use the filter (compiled as a plugin), through the `h5repack` command line tool. As I mentioned in [this comment](https://github.com/Blosc/hdf5-blosc/issues/11#issuecomment-240444100) I can use the filter fine from h5py by passing a 7 element tuple with filter options.

However, with `h5repack`, I run into trouble if I try to do the same.

First I create a test file with:

```
$ python
Python 3.5.1+ (default, Mar 30 2016, 22:46:26) 
[GCC 5.3.1 20160330] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import h5py
>>> with h5py.File('test.hdf5', 'w') as outfile:
...     outfile.create_dataset('test', (3, 3), data=[1, 2, 3, 4, 5, 6, 7, 8, 9])
... 
<HDF5 dataset "test": shape (3, 3), type "<i8">
>>> 
```

The result is a file with an uncompressed dataset:

```
$ h5dump -p test.hdf5    
HDF5 "test.hdf5" {
GROUP "/" {
   DATASET "test" {
      DATATYPE  H5T_STD_I64LE
      DATASPACE  SIMPLE { ( 3, 3 ) / ( 3, 3 ) }
      STORAGE_LAYOUT {
         CONTIGUOUS
         SIZE 72
         OFFSET 2144
      }
      FILTERS {
         NONE
      }
      FILLVALUE {
         FILL_TIME H5D_FILL_TIME_IFSET
         VALUE  0
      }
      ALLOCATION_TIME {
         H5D_ALLOC_TIME_LATE
      }
      DATA {
      (0,0): 1, 2, 3,
      (1,0): 4, 5, 6,
      (2,0): 7, 8, 9
      }
   }
}
}
```

I then try to repack it, applying the Blosc filter:

```
$ h5repack -f UD=32001,7,2,2,8,72,9,1,2 test.hdf5 test_blosc.hdf5
h5repack error: incorrect number of compression parameters in <UD=32001,7,2,2,8,72,9,1,2>
```

That's strange, I'm giving 7 as the element count, followed by 7 elements, which I believe have the following meaning: FILTER_BLOSC_VERSION (2 in my case), BLOSC_VERSION_FORMAT (2 in my case), datatype size (8 in my case), output buffer size (using input buffer size here, 72 bytes), compression level (using 9 here), shuffle filter (using byte-based shuffling, so 1) and finally the compressor (using BLOSC_LZ4HC here, so 2).

First suspecting some parsing bug in the h5repack tool, I changed the leading 7 (the element count) to 8 instead, and then it seems to accept the command line, but instead I get this error:

```
$ h5repack -f UD=32001,8,2,2,8,72,9,1,2 test.hdf5 test_blosc.hdf5
`clevel` parameter must be between 0 and 9!
`clevel` parameter must be between 0 and 9!
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 139840359859968:
  #000: ../../../src/H5D.c line 426 in H5Dclose(): can't decrement count on dataset ID
    major: Dataset
    minor: Unable to decrement reference count
  #001: ../../../src/H5I.c line 1549 in H5I_dec_app_ref_always_close(): can't decrement ID ref count
    major: Object atom
    minor: Unable to decrement reference count
  #002: ../../../src/H5I.c line 1491 in H5I_dec_app_ref(): can't decrement ID ref count
    major: Object atom
    minor: Unable to decrement reference count
  #003: ../../../src/H5Dint.c line 1438 in H5D_close(): unable to destroy chunk cache
    major: Dataset
    minor: Unable to release object
  #004: ../../../src/H5Dchunk.c line 5164 in H5D__chunk_dest(): unable to flush one or more raw data chunks
    major: Low-level I/O
    minor: Unable to flush data from cache
  #005: ../../../src/H5Dchunk.c line 2641 in H5D__chunk_cache_evict(): cannot flush indexed storage buffer
    major: Low-level I/O
    minor: Write failed
  #006: ../../../src/H5Dchunk.c line 2528 in H5D__chunk_flush_entry(): output pipeline failed
    major: Data filters
    minor: Filter operation failed
  #007: ../../../src/H5Z.c line 1412 in H5Z_pipeline(): filter returned failure
    major: Data filters
    minor: Write failed
  #008: src/blosc_filter.c line 229 in blosc_filter(): Blosc compression error
    major: Data filters
    minor: Callback failed
  #009: ../../../src/H5Dint.c line 1401 in H5D_close(): unable to flush cached dataset info
    major: Dataset
    minor: Write failed
  #010: ../../../src/H5Dint.c line 2329 in H5D__flush_real(): unable to flush raw data
    major: Dataset
    minor: Unable to flush data from cache
  #011: ../../../src/H5Dchunk.c line 2112 in H5D__chunk_flush(): unable to flush one or more raw data chunks
    major: Dataset
    minor: Unable to flush data from cache
  #012: ../../../src/H5Dchunk.c line 2528 in H5D__chunk_flush_entry(): output pipeline failed
    major: Data filters
    minor: Filter operation failed
  #013: ../../../src/H5Z.c line 1412 in H5Z_pipeline(): filter returned failure
    major: Data filters
    minor: Write failed
  #014: src/blosc_filter.c line 229 in blosc_filter(): Blosc compression error
    major: Data filters
    minor: Callback failed
h5repack error: <test.hdf5>: Could not copy data to: test_blosc.hdf5
```

The two <pre>`clevel` parameter must be between 0 and 9!</pre> printouts are from the filter itself (is it taking the 72 as the compression level..?).

...so, I'm puzzled. What is the correct way to use the filter through `h5repack`?

I looked a bit at the filter code, and in `blosc_set_local(...)`, a values array of length 8 seems to be used, while in `blosc_filter(...)` only 7 elements are used. Could this discrepancy have something to do with it?

I tried changing my command line to

```
$ h5repack -f UD=32001,8,2,2,8,72,9,1,2,0 test.hdf5 test_blosc.hdf5
```

E.g. declaring 8 elements and adding a 0 at the end, and also

```
$ h5repack -f UD=32001,8,0,2,2,8,72,9,1,2 test.hdf5 test_blosc.hdf5
```

E.g. the same but adding a 0 at the beginning. But in both cases I get that `h5repack error: incorrect number of compression parameters [...]` error from the tool. So I'm guessing it is some sort of parsing bug in the `h5repack` tool?

**Comment by estan** [comment]
Looking at [this page](https://github.com/HDFGroup/datacontainer/wiki/Big-Deep-HDF), it seems that someone has gotten it to work. The example there uses:

```
$ h5repack -L -l CHUNK=45x180 -f UD=32001,2,4,8 ~/TES-Aura_L3-CH4_r0000010410_F01_07.he5 test.he5
```

So just two parameters (4 and 8).

But if I test something similar on my test file, I get:

```
$ h5repack -l CHUNK=3x3 -f UD=32001,2,4,8 test.hdf5 test_blosc.hdf5 -h
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 139651109697280:
  #000: ../../../src/H5D.c line 426 in H5Dclose(): can't decrement count on dataset ID
    major: Dataset
    minor: Unable to decrement reference count
  #001: ../../../src/H5I.c line 1549 in H5I_dec_app_ref_always_close(): can't decrement ID ref count
    major: Object atom
    minor: Unable to decrement reference count
  #002: ../../../src/H5I.c line 1491 in H5I_dec_app_ref(): can't decrement ID ref count
    major: Object atom
    minor: Unable to decrement reference count
  #003: ../../../src/H5Dint.c line 1438 in H5D_close(): unable to destroy chunk cache
    major: Dataset
    minor: Unable to release object
  #004: ../../../src/H5Dchunk.c line 5164 in H5D__chunk_dest(): unable to flush one or more raw data chunks
    major: Low-level I/O
    minor: Unable to flush data from cache
  #005: ../../../src/H5Dchunk.c line 2641 in H5D__chunk_cache_evict(): cannot flush indexed storage buffer
    major: Low-level I/O
    minor: Write failed
  #006: ../../../src/H5Dchunk.c line 2528 in H5D__chunk_flush_entry(): output pipeline failed
    major: Data filters
    minor: Filter operation failed
  #007: ../../../src/H5Z.c line 1412 in H5Z_pipeline(): filter returned failure
    major: Data filters
    minor: Write failed
  #008: ../../../src/H5Dint.c line 1401 in H5D_close(): unable to flush cached dataset info
    major: Dataset
    minor: Write failed
  #009: ../../../src/H5Dint.c line 2329 in H5D__flush_real(): unable to flush raw data
    major: Dataset
    minor: Unable to flush data from cache
  #010: ../../../src/H5Dchunk.c line 2112 in H5D__chunk_flush(): unable to flush one or more raw data chunks
    major: Dataset
    minor: Unable to flush data from cache
  #011: ../../../src/H5Dchunk.c line 2528 in H5D__chunk_flush_entry(): output pipeline failed
    major: Data filters
    minor: Filter operation failed
  #012: ../../../src/H5Z.c line 1412 in H5Z_pipeline(): filter returned failure
    major: Data filters
    minor: Write failed
h5repack error: <test.hdf5>: Could not copy data to: test_blosc.hdf5
```

The filter is returning an error somewhere, but not printing anything.. Any ideas?

**Comment by estan** [comment]
Okay, this last error (the one without a printout from the filter) was actually due to me using such a small test file. With a 3x3x3 dataset instead, which can actually be compressed into a size less than the input size, it succeeds.

However, the strangeness with the `h5repack` argument parsing remains. I'm not quite sure what is going on, but if I add a printout of the elements in `cd_values` in `blosc_filter(...)`, I get the following strange behavior:

```
$ h5repack -l CHUNK=3x3x3 -f UD=32001,7,0,0,0,0,0,9,1 test.hdf5 test_blosc.hdf5
2 2 8 216 9 1 1
$
```

but

```
$ h5repack -l CHUNK=3x3x3 -f UD=32001,7,0,0,0,0,0,9,2 test.hdf5 test_blosc.hdf5
2 2 8 216 9 2 2
```

Note how I only changed a single 1 to a 2 in the second run, but both `cd_values[5]` and `cd_values[6]` was changed (!). So something strange is definitely weird.

**Comment by estan** [comment]
And also:

```
$ h5repack -l CHUNK=3x3x3 -f UD=32001,7,0,0,0,0,9,1,1 test.hdf5 test_blosc.hdf5
2 2 8 216 1 1 1
```

Very puzzling..

**Comment by FrancescAlted** [comment]
Not sure what's going on here, but being a h5repack issue, I wonder whether @epourmal can shed some light.

**Comment by epourmal** [comment]
According to the error stack the BLOSC filter failed to compress a chunk. 

I would first check if you can create a compressed dataset with a simple C or h5py program.
Here is an example how to use external dynamically loaded filter using BZIP2 https://svn.hdfgroup.org/hdf5_plugins/trunk/BZIP2/example/h5ex_d_bzip2.c

We also have BLOSC example, but I am not sure that we ever made it to work https://svn.hdfgroup.org/hdf5_plugins/trunk/BLOSC/example/h5ex_d_blosc.c.

**Comment by estan** [comment]
@epourmal Hi, and thanks for chiming in.

I debugged the failure I mentioned in [this comment](https://github.com/Blosc/hdf5-blosc/issues/13#issuecomment-240703844) and found out that it was due to me using such a small (3x3) dataset for testing. The Blosc filter failed since it could not compress into a size that would be < the input size (because of overhead). When I changed to using a slightly larger (3x3x3) test dataset, the compression succeeded. So the reason for that failure is explained now. Creating a Blosc-compressed dataset using this filter from `h5py` also works fine, by passing 32001 as filter number and a 7-tuple containing the filter parameters to `create_dataset`.

The problem I have now has to do with the `h5repack` command and how the `UD=...` command line parameter is treated/parsed and fed to the the filter functions. The printouts I made [here](https://github.com/Blosc/hdf5-blosc/issues/13#issuecomment-240735142) and [here](https://github.com/Blosc/hdf5-blosc/issues/13#issuecomment-240735619) were made from the `cd_values` array in the `blosc_filter` function in the filter, which is called by the HDF5 library.

I think (but I'm not sure) that this strange behavior could be some command line parsing bug in `h5repack`? I've looked a little at the parsing code there, but it's a bit complicated so I haven't gotten anywhere.

In the meantime, it would be interesting if someone else could test and see if they can reproduce the issue. [Here](https://github.com/Blosc/hdf5-blosc/files/426705/testcase.tar.gz) is a minimal test case to try it out, which you can run as follows (my output shown):

```
estan@newton:~/testcase$ make
gcc -O2 src/blosc_filter.c src/blosc_plugin.c -fPIC -shared `pkg-config --cflags --libs hdf5` -lblosc -o libblosc_filter.so
```

This builds the filter as a plugin, the code in `src/` in the testcase was taken from this GitHub project. The only change I've done is to add

``` c
    /* DEBUG: Print the cd_values array. */
    for (size_t i = 0; i < cd_nelmts; ++i)
        fprintf(stderr, "%u ", cd_values[i]);
    fprintf(stderr, "\n");
```

to the `blosc_filter` function in `src/blosc_filter.c`, to print the `cd_values` array.

And then:

```
estan@newton:~/testcase$ HDF5_PLUGIN_PATH=. h5repack -l CHUNK=3x3x3 -f UD=32001,7,0,0,0,0,0,9,1 input.hdf5 output.hdf5
2 2 8 216 9 1 1 
```

And:

```
estan@newton:~/testcase$ HDF5_PLUGIN_PATH=. h5repack -l CHUNK=3x3x3 -f UD=32001,7,0,0,0,0,0,9,2 input.hdf5 output.hdf5
2 2 8 216 9 2 2 
2 2 8 216 9 2 2 
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 140211060188928:
  #000: ../../../src/H5D.c line 426 in H5Dclose(): can't decrement count on dataset ID
    major: Dataset
    minor: Unable to decrement reference count
  #001: ../../../src/H5I.c line 1549 in H5I_dec_app_ref_always_close(): can't decrement ID ref count
    major: Object atom
    minor: Unable to decrement reference count
  #002: ../../../src/H5I.c line 1491 in H5I_dec_app_ref(): can't decrement ID ref count
    major: Object atom
    minor: Unable to decrement reference count
  #003: ../../../src/H5Dint.c line 1438 in H5D_close(): unable to destroy chunk cache
    major: Dataset
    minor: Unable to release object
  #004: ../../../src/H5Dchunk.c line 5164 in H5D__chunk_dest(): unable to flush one or more raw data chunks
    major: Low-level I/O
    minor: Unable to flush data from cache
  #005: ../../../src/H5Dchunk.c line 2641 in H5D__chunk_cache_evict(): cannot flush indexed storage buffer
    major: Low-level I/O
    minor: Write failed
  #006: ../../../src/H5Dchunk.c line 2528 in H5D__chunk_flush_entry(): output pipeline failed
    major: Data filters
    minor: Filter operation failed
  #007: ../../../src/H5Z.c line 1412 in H5Z_pipeline(): filter returned failure
    major: Data filters
    minor: Write failed
  #008: src/blosc_filter.c line 180 in blosc_filter(): this Blosc library version does not have support for the bitshuffle filter.  Please update to >= 1.8
    major: Data filters
    minor: Callback failed
  #009: ../../../src/H5Dint.c line 1401 in H5D_close(): unable to flush cached dataset info
    major: Dataset
    minor: Write failed
  #010: ../../../src/H5Dint.c line 2329 in H5D__flush_real(): unable to flush raw data
    major: Dataset
    minor: Unable to flush data from cache
  #011: ../../../src/H5Dchunk.c line 2112 in H5D__chunk_flush(): unable to flush one or more raw data chunks
    major: Dataset
    minor: Unable to flush data from cache
  #012: ../../../src/H5Dchunk.c line 2528 in H5D__chunk_flush_entry(): output pipeline failed
    major: Data filters
    minor: Filter operation failed
  #013: ../../../src/H5Z.c line 1412 in H5Z_pipeline(): filter returned failure
    major: Data filters
    minor: Write failed
  #014: src/blosc_filter.c line 180 in blosc_filter(): this Blosc library version does not have support for the bitshuffle filter.  Please update to >= 1.8
    major: Data filters
    minor: Callback failed
h5repack error: <input.hdf5>: Could not copy data to: output.hdf5
(analysis) estan@newton:~/testcase$ HDF5_PLUGIN_PATH=. h5repack -l CHUNK=3x3x3 -f UD=32001,7,0,0,0,0,9,1,1 input.hdf5 output.hdf5
2 2 8 216 1 1 1 
```

And also:

```
estan@newton:~/testcase$ HDF5_PLUGIN_PATH=. h5repack -l CHUNK=3x3x3 -f UD=32001,7,0,0,0,0,9,1,1 input.hdf5 output.hdf5
2 2 8 216 1 1 1
```

In all three of these cases, what I find strange is the effect of the input `UD=...` command line parameter has on the `cd_values`. It is expected that the filter overwrites the first four elements (it does so in its initialization function), but what about the remaining three? I can't find an explanation for why they don't correspond to my input :(

Much appreciated if someone could try this out on their system.

**Comment by estan** [comment]
What's also interesting is creating the same dataset from scratch using `h5py`, and looking at the printout of `cd_values`:

_test.py_:

``` python
import h5py

with h5py.File('output.hdf5', 'w') as outfile:
    outfile.create_dataset(
        'test',
        shape=(3, 3, 3),
        data=[
            1, 2, 3, 4, 5, 6, 7, 8, 9,
            1, 2, 3, 4, 5, 6, 7, 8, 9,
            1, 2, 3, 4, 5, 6, 7, 8, 9,
        ],
        shuffle=False,
        compression=32001,
        compression_opts=(0, 0, 0, 0, 9, 1, 1)
    )
```

```
estan@newton:~/testcase$ HDF5_PLUGIN_PATH=. python test.py 
2 2 8 216 9 1 1
```

Note how, despite the same input filter options (`0, 0, 0, 0, 9, 1, 1`) as in the third `h5repack` invokation in my previous comment, the output is different (and as expected) here with `h5py`, but not with `h5repack`.

So this further makes me believe there's something going wrong with the command line parsing in `h5repack`?

The result (`output.hdf5`) from the above `h5py` program also looks fine:

```
estan@newton:~/testcase$ HDF5_PLUGIN_PATH=. h5dump -p output.hdf5 
HDF5 "output.hdf5" {
GROUP "/" {
   DATASET "test" {
      DATATYPE  H5T_STD_I64LE
      DATASPACE  SIMPLE { ( 3, 3, 3 ) / ( 3, 3, 3 ) }
      STORAGE_LAYOUT {
         CHUNKED ( 3, 3, 3 )
         SIZE 47 (4.596:1 COMPRESSION)
      }
      FILTERS {
         USER_DEFINED_FILTER {
            FILTER_ID 32001
            COMMENT blosc
            PARAMS { 2 2 8 216 9 1 1 }
         }
      }
      FILLVALUE {
         FILL_TIME H5D_FILL_TIME_ALLOC
         VALUE  0
      }
      ALLOCATION_TIME {
         H5D_ALLOC_TIME_INCR
      }
2 2 8 216 9 1 1 
      DATA {
      (0,0,0): 1, 2, 3,
      (0,1,0): 4, 5, 6,
      (0,2,0): 7, 8, 9,
      (1,0,0): 1, 2, 3,
      (1,1,0): 4, 5, 6,
      (1,2,0): 7, 8, 9,
      (2,0,0): 1, 2, 3,
      (2,1,0): 4, 5, 6,
      (2,2,0): 7, 8, 9
      }
   }
}
}
```

**Comment by estan** [comment]
So to summarize:

With `h5py`:

```
estan@newton:~/testcase$ HDF5_PLUGIN_PATH=. python test.py 
2 2 8 216 9 1 1
```

With `h5repack`:

```
estan@newton:~/testcase$ HDF5_PLUGIN_PATH=. h5repack -l CHUNK=3x3x3 -f UD=32001,7,0,0,0,0,9,1,1 input.hdf5 output.hdf5
2 2 8 216 1 1 1
```

Why does `cd_values` become `2 2 8 216 1 1 1` in the `h5repack` case, but `2 2 8 216 9 1 1` (the expected) in the `h5py` case?

**Comment by estan** [comment]
I've now made a minimal noop plugin, which also shows the issue, and made [a post](https://lists.hdfgroup.org/pipermail/hdf-forum_lists.hdfgroup.org/2016-August/009761.html) about it to the hdf-forum mailing list. Let's hope someone at HDF Group responds there and can confirm the bug. I'll keep this issue open for a little while until someone responds there, if that's okay.

**Comment by estan** [comment]
A second user has now confirmed the bug in that thread, so I think I'll close this issue now since it's not related to the Blosc filter. If you're interested you can follow the thread instead. Thanks @FrancescAlted and @epourmal.

**Comment by estan** [comment]
Just as a follow-up: Allen Byrne from HDFGroup has now replied with a patch for h5repack to the thread linked above. I've tried the patch and it seems to work fine.

**Comment by epourmal** [comment]
Thank you, Elvis!

I talked to Allen and we discovered that his fix that was created two years? ago, somehow fell through the cracks and never made into the releases. That is why I thought that it is a filter problem in the first place.

Elena
On Aug 25, 2016, at 6:04 PM, Elvis Stansvik <notifications@github.com<mailto:notifications@github.com>> wrote:

Just as a follow-up: Allen Byrne from HDFGroup has now replied with a patch for h5repack to the thread linked above. I've tried the patch and it seems to work fine.
